File: two_pointers/remove_spaces.py
def remove_leading_spaces(arr):
    slow = 0
    fast = 0

    while fast < len(arr):
        if arr[fast] != ' ' or (arr[fast] == ' ' and (slow == 0 or arr[slow - 1] != ' ')):
            arr[slow] = arr[fast]
            slow += 1
        fast += 1

    return arr[:slow]

result = remove_leading_spaces([
  " ",
  " ",
  "H",
  "i",
  " ",
  " ",
  " ",
  " ",
  "!",
  "!",
  " ",
  " ",
]); # [" ", "H", "i", " ", "!", "!", " "];
print(result)

File: two_pointers/remove_spaces.js
function remove_leading_spaces(arr) {
  let slowPointer = 0;
  let fastPointer = 0;

  while (fastPointer < arr.length) {
    if (
      arr[fastPointer] !== " " ||
      (arr[fastPointer] === " " &&
        (slowPointer === 0 || arr[slowPointer - 1] !== " "))
    ) {
      arr[slowPointer] = arr[fastPointer];
      slowPointer++;
    }
    fastPointer++;
  }
  return arr.slice(0, slowPointer);
}

const result = remove_leading_spaces([
  " ",
  " ",
  "H",
  "i",
  " ",
  " ",
  " ",
  " ",
  "!",
  "!",
  " ",
  " ",
]); // [" ", "H", "i", " ", "!", "!", " "];
console.log(result);
